### PR TITLE
Convert Naming Styles to a grid-based UI

### LINKED ISSUE #14358: Changes to "Naming Styles" aren't persisted
**Version Used**: 

**Steps to Reproduce**:

1. Create a naming style called "_camelCased" with _ as a required prefix, but *Pascal Cased* for capitalization.
2. Go back to Tools Options and change the capitalization to camelCased.

**Expected Behavior**:
The rule should be updated.

**Actual Behavior**:
It isn't.

### PR DESCRIPTION
**Pull request by dpoeschl**
This change also introduces Specification and Style management (deleting and editing).

Also, fixes #14358

### CODE REVIEW

**Comment by dpoeschl** [comment]
Here's the proposed UI (subject to probable change in the future):

image

image

image

image

image

**Comment by dpoeschl** [comment]
Tagging @dotnet/roslyn-ide for review.

**Comment by CyrusNajmabadi** [comment]
image

What is the blue line indicating?

Why does one of these not have the delete box enabled?

File: src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingRule.cs [hunk 1]
```diff
@@ -1,39 +1,25 @@
...
         {
-            return SymbolSpecification.AppliesTo(symbol, categorizationService);
-        }
-
-        internal NamingRule GetBestMatchingRule(ISymbol symbol, ISymbolCategorizationService categorizationService)
-        {
-            Debug.Assert(SymbolSpecification.AppliesTo(symbol, categorizationService));
-            var matchingChild = Children.FirstOrDefault(r => r.AppliesTo(symbol, categorizationService));
-            return matchingChild?.GetBestMatchingRule(symbol, categorizationService) ?? this;
+            return SymbolSpecification.AppliesTo(symbol);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingRule.cs]: =>

**Comment by dpoeschl** [comment]
@CyrusNajmabadi That's the "edit" glyph. Perhaps we can find a better one.

Delete is disabled on specifications/styles that are in use by existing rules.

File: src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingStylePreferencesInfo.cs [hunk 2]
```diff
@@ -30,15 +28,8 @@ internal bool TryGetApplicableRule(ISymbol symbol, ISymbolCategorizationService
...
-            var matchingRule = NamingRules.FirstOrDefault(r => r.AppliesTo(symbol, categorizationService));
-            if (matchingRule == null)
-            {
-                applicableRule = null;
-                return false;
-            }
-
-            applicableRule = matchingRule.GetBestMatchingRule(symbol, categorizationService);
-            return true;
+            applicableRule = NamingRules.FirstOrDefault(r => r.AppliesTo(symbol));
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingStylePreferencesInfo.cs]: Does this run on every symbol?  If so can we avoid allocating?

File: src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingStylePreferencesInfo.cs [hunk 3]
```diff
@@ -49,6 +40,12 @@ private bool IsSymbolNameAnalyzable(ISymbol symbol)
                 return false;
             }
 
+            var propertySymbol = symbol as IPropertySymbol;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingStylePreferencesInfo.cs]: If this runs on every symbol we should attempt .SymbolKind == checks first.  'is/as' are expensive and do add up.

File: src/Features/Core/Portable/FeaturesResources.resx [hunk 4]
```diff
@@ -1082,4 +1082,87 @@ This version used in: {2}</value>
...
+    <value>All lowercase</value>
+  </data>
+  <data name="All_uppercase" xml:space="preserve">
+    <value>All uppercase</value>
+  </data>
+  <data name="Begins_with_0" xml:space="preserve">
+    <value>Begins with '{0}'</value>
+  </data>
+  <data name="Begins_with_0_camel_case" xml:space="preserve">
+    <value>Begins with '{0}'; camel case</value>
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/FeaturesResources.resx]: Hrmm..  should we instead generate something like a comma/semicolon separated list dynamically?  I'm not sure how that would work with loc.  but i worry that you'd have to create resource strings for these sort of combined rules.

File: src/VisualStudio/CSharp/Impl/Options/NamingStylesOptionPage.cs [hunk 5]
```diff
@@ -1,19 +1,35 @@
...
 using Microsoft.VisualStudio.LanguageServices.Implementation.Options;
-using Microsoft.VisualStudio.LanguageServices.Implementation.Options.Style.NamingPreferences;
+using Microsoft.VisualStudio.LanguageServices.Implementation.Options.Style;
 
 namespace Microsoft.VisualStudio.LanguageServices.CSharp.Options
 {
     [Guid(Guids.CSharpOptionPageNamingStyleIdString)]
     internal class NamingStylesOptionPage : AbstractOptionPage
     {
+        private NamingStyleOptionPageControl _grid;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/CSharp/Impl/Options/NamingStylesOptionPage.cs]: readonly?

File: src/VisualStudio/Core/Def/ServicesVSResources.resx [hunk 6]
```diff
@@ -762,4 +762,25 @@ Additional information: {1}</value>
   <data name="Prefer_inlined_variable_declaration" xml:space="preserve">
     <value>Prefer inlined variable declaration</value>
   </data>
+  <data name="Some_naming_rules_are_incomplete" xml:space="preserve">
+    <value>Some naming rules are incomplete. Please complete or remove them.</value>
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/ServicesVSResources.resx]: key doesn't match value.

File: src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/NamingStyleOptionPageControl.xaml.cs [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Options.Style
+{
+    internal partial class NamingStyleOptionPageControl : AbstractOptionPageControl
+    {
+        private NamingStyleOptionPageViewModel _viewModel;
+        private readonly string _languageName;
+        private readonly INotificationService _notificationService;
+
+        private readonly EnforcementLevel[] _notifications = new[]
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/NamingStyleOptionPageControl.xaml.cs]: not a fan of raw arrays.  Consider an ImmutableArray.

File: src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/NamingStyleOptionPageViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+                n.ID,
+                n.ItemName,
+                n.SymbolKindList.Where(s => s.IsChecked).Select(k => k.CreateSymbolKindOrTypeKind()).ToList(),
+                n.AccessibilityList.Where(s => s.IsChecked).Select(a => new SymbolSpecification.AccessibilityKind(a._accessibility)).ToList(),
+                n.ModifierList.Where(s => s.IsChecked).Select(m => new SymbolSpecification.ModifierKind(m._modifier)).ToList()));
+
+            Specifications.Clear();
+            foreach (var specification in symbolSpecifications)
+            {
+                Specifications.Add(specification);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/NamingStyleOptionPageViewModel.cs]: AddRange?

File: src/Features/Core/Portable/FeaturesResources.resx (hunk 4 above)
Comment (dpoeschl) [inline src/Features/Core/Portable/FeaturesResources.resx]: I believe these can be deleted. They're leftover from the predefined options stuff @rchande did. Will remove.

File: src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/NamingStyleOptionPageViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+            }
+
+            foreach (var rule in CodeStyleItems)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/NamingStyleOptionPageViewModel.cs]: can you doc what this code is doing.  it's not clear to me.

**Comment by CyrusNajmabadi** [comment]
👍

File: src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/NamingStyleOptionPageViewModel.cs (hunk 9 above)
Comment (dpoeschl) [inline src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/NamingStyleOptionPageViewModel.cs]: Yep

File: src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/Serialization/SymbolSpecification.cs [hunk 10]
```diff
@@ -199,10 +180,10 @@ private void PopulateModifierListFromXElement(XElement modifierListElement)
 
         private void PopulateCustomTagListFromXElement(XElement customTagListElement)
         {
-            RequiredCustomTagList = new List<string>();
+            var requiredCustomTagList = new List<string>();
             foreach (var customTag in customTagListElement.Elements("CustomTag"))
             {
```
Comment (Pilchie) [inline src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/Serialization/SymbolSpecification.cs]: Does this need to assign to `RequiredCustomTagList`?

File: src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/SymbolSpecification/SymbolSpecificationDialog.xaml [hunk 11]
```diff
@@ -144,7 +144,7 @@
                     <RowDefinition Height="*"/>
                     <RowDefinition Height="Auto" />
                 </Grid.RowDefinitions>
-                <Label Grid.Row="0" Content="{Binding ElementName=dialog, Path=CustomTagsLabelText}"/>
+                <!--<Label Grid.Row="0" Content="{Binding ElementName=dialog, Path=CustomTagsLabelText}"/>
```
Comment (Pilchie) [inline src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/SymbolSpecification/SymbolSpecificationDialog.xaml]: Just remove commented things?

File: src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/SymbolSpecification/SymbolSpecificationDialog.xaml.cs [hunk 12]
```diff
@@ -111,25 +111,9 @@ private void DeselectAllModifiers(object sender, RoutedEventArgs e)
...
-            foreach (var item in CustomTags.Items.OfType<SymbolSpecificationViewModel.CustomTagViewModel>())
-            {
-                item.IsChecked = false;
-            }
-        }
-
         private void OK_Click(object sender, RoutedEventArgs e)
         {
-            if (_viewModel.TrySubmit())
+            // if (_viewModel.TrySubmit())
```
Comment (Pilchie) [inline src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/SymbolSpecification/SymbolSpecificationDialog.xaml.cs]: Remove commented?

File: src/VisualStudio/VisualBasic/Impl/Options/NamingStylesOptionPage.vb [hunk 13]
```diff
@@ -1,15 +1,31 @@
...
+
         Protected Overrides Function CreateOptionPage(serviceProvider As IServiceProvider) As AbstractOptionPageControl
-            Return New NamingStylesOptionPageControl(serviceProvider, LanguageNames.VisualBasic)
+            _grid = New NamingStyleOptionPageControl(serviceProvider, LanguageNames.VisualBasic)
+            Return _grid
         End Function
+
+        Protected Overrides Sub OnApply(e As PageApplyEventArgs)
+            If _grid.ContainsErrors() Then
+                MessageBox.Show(ServicesVSResources.Some_naming_rules_are_incomplete)
```
Comment (Pilchie) [inline src/VisualStudio/VisualBasic/Impl/Options/NamingStylesOptionPage.vb]: This should probably use the VS message box service.

File: src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/Serialization/SymbolSpecification.cs (hunk 10 above)
Comment (dpoeschl) [inline src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/Serialization/SymbolSpecification.cs]: This method should be removed. There are no more custom tags. :cry:

**Comment by kuhlenh** [comment]
@dpoeschl we have "Style" in this UI but "Preference" in the General one. Should we make them match by renaming one or the other?

File: src/Features/Core/Portable/FeaturesResources.resx (hunk 4 above)
Comment (dpoeschl) [inline src/Features/Core/Portable/FeaturesResources.resx]: But I do have to loc the new gigantic default option string...

File: src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingRule.cs (hunk 1 above)
Comment (dpoeschl) [inline src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingRule.cs]: Done.

File: src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingStylePreferencesInfo.cs (hunk 2 above)
Comment (dpoeschl) [inline src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingStylePreferencesInfo.cs]: Yes, will fix.

File: src/VisualStudio/Core/Def/ServicesVSResources.resx (hunk 6 above)
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/ServicesVSResources.resx]: Done

File: src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/NamingStyleOptionPageControl.xaml.cs (hunk 7 above)
Comment (dpoeschl) [inline src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/NamingStyleOptionPageControl.xaml.cs]: Done

File: src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/NamingStyleOptionPageViewModel.cs (hunk 8 above)
Comment (dpoeschl) [inline src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/NamingStyleOptionPageViewModel.cs]: ObservableCollection doesn't have AddRange.

File: src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/Serialization/SymbolSpecification.cs (hunk 10 above)
Comment (dpoeschl) [inline src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/Serialization/SymbolSpecification.cs]: All references to the CustomTagList (this and others) are now removed (once I push)

File: src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/SymbolSpecification/SymbolSpecificationDialog.xaml (hunk 11 above)
Comment (dpoeschl) [inline src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/SymbolSpecification/SymbolSpecificationDialog.xaml]: Done

File: src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/SymbolSpecification/SymbolSpecificationDialog.xaml.cs (hunk 12 above)
Comment (dpoeschl) [inline src/VisualStudio/Core/Impl/Options/Style/NamingPreferences/SymbolSpecification/SymbolSpecificationDialog.xaml.cs]: Actually, uncomment and do the right thing.

File: src/VisualStudio/VisualBasic/Impl/Options/NamingStylesOptionPage.vb (hunk 13 above)
Comment (dpoeschl) [inline src/VisualStudio/VisualBasic/Impl/Options/NamingStylesOptionPage.vb]: Fixed up.

**Comment by dpoeschl** [comment]
@kuhlenh If you're talking about this versus the general code styles page, I'd argue that they're different. The general code styles page has options where you're picking what you prefer in a general sense. Here, it's still a "naming preference" but I think "Style" describes it more accurately. But I could be swayed.

**Comment by dpoeschl** [comment]
@CyrusNajmabadi You're right, the icon looks awful because it's being truncated. I'll see what I can do.

**Comment by dpoeschl** [comment]
@CyrusNajmabadi Fixed it. Looking much better.

image

**Comment by dpoeschl** [comment]
@Pilchie @CyrusNajmabadi (@dotnet/roslyn-ide) I've updated the PR and responded to all of the feedback. Can you take a second look?

**Comment by dpoeschl** [comment]
Tagging @ManishJayaswal and/or @MattGertz for RC ask mode approval

**Comment by ManishJayaswal** [comment]
Approved pending CRs and test runs.

**Comment by dpoeschl** [comment]
retest windows_debug_unit64_prtest please
Failure: https://ci.dot.net/job/dotnet_roslyn/job/dev15-rc/job/windows_debug_unit64_prtest/130/

**Comment by dpoeschl** [comment]
retest windows_debug_unit64_prtest please

**Pull request merged**